src/main/java/com/nvidia/spark/rapids/jni/GpuTimeZoneDB.java

```diff
@@ -109,23 +122,94 @@ private static synchronized void closeResources()  {
+    if (transitions != null) {
+      transitions.close();
+      transitions = null;
     }
   }
 
-  public static ColumnVector fromTimestampToUtcTimestamp(ColumnVector input, ZoneId currentTimeZone) {
-    // TODO: Remove this check when all timezones are supported
-    // (See https://github.com/NVIDIA/spark-rapids/issues/6840)
-    if (!isSupportedTimeZone(currentTimeZone)) {
-      throw new IllegalArgumentException(String.format("Unsupported timezone: %s",
-          currentTimeZone.toString()));
+  private static long getScalefactor(ColumnVector input){
+    DType inputType = input.getType();
+    if (inputType == DType.TIMESTAMP_SECONDS){
+      return 1;
+    } else if (inputType == DType.TIMESTAMP_MILLISECONDS){
+      return 1000;
+    } else if (inputType == DType.TIMESTAMP_MICROSECONDS){
+      return 1000*1000;
+    }
+    throw new UnsupportedOperationException("Input Type is not of TIMESTAMP type");
+  }
+
+  public static boolean isSupportedTimeZone(String zoneId) {
+    try {
+      getZoneId(zoneId); // check that zoneId is a valid zone
+      return true;
+    } catch (ZoneRulesException e) {
+      return false;
+    }
+  }
+
+  // enforce that all timestamps, regardless of timezone, be less than the desired date
+  private static boolean isValidInput(ColumnVector input, ZoneId zoneId){
+    if (zoneId.getRules().isFixedOffset()){
+      return true;
+    }
+    boolean isValid = false;
+    long scaleFactor = getScalefactor(input);
+    try (Scalar targetTimestamp = Scalar.timestampFromLong(input.getType(), maxTimestamp*scaleFactor);
+         ColumnVector compareCv = input.binaryOp(BinaryOp.GREATER, targetTimestamp, DType.BOOL8);
+         Scalar isGreater = compareCv.any() ) {
+      if (!isGreater.isValid()) isValid = true;
+      else isValid = !isGreater.getBoolean();
+    } catch (Exception e) {
+      log.error("Error validating input timestamps", e);
+      // don't need to throw error, can try CPU processing
+      return false;
+    }
+    return isValid;
+  }
+
+  private static ColumnVector cpuChangeTimestampTz(ColumnVector input, ZoneId currentTimeZone, ZoneId targetTimeZone) {
+    ColumnVector resultCV = null;
+    try (HostColumnVector hostCV = input.copyToHost()) {
+      // assuming we don't have more than 2^31-1 rows
+      int rows = (int) hostCV.getRowCount();
+      DType inputType = input.getType();
+      long scaleFactor = getScalefactor(input);
+      
+      try (HostColumnVector.Builder builder = HostColumnVector.builder(inputType, rows)) {
+        for (int i = 0; i < rows; i++) {
+          if (hostCV.isNull(i)) {
+            builder.appendNull();
+            continue;
+          }
+          
+          long timestamp = hostCV.getLong(i);
+          long unitOffset = timestamp % scaleFactor;
+          timestamp /= scaleFactor;
+          Instant instant = Instant.ofEpochSecond(timestamp);
+          timestamp = instant.atZone(targetTimeZone).toLocalDateTime().atZone(currentTimeZone)
```

Comment:
Could you please add a comment explaining this conversion? It would help people to understand this code.